realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTests.java

```diff
@@ -996,4 +999,187 @@ public void realmListRemoveAllFromRealm() {
+        Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                DynamicRealm realm = DynamicRealm.getInstance(defaultConfig);
+                openedLatch.countDown();
+                result1.set(realm.waitForChange());
+                count1.set(realm.where(AllTypes.CLASS_NAME).count());
+                changedLatch.countDown();
+                result2.set(realm.waitForChange());
+                count2.set(realm.where(AnnotationIndexTypes.CLASS_NAME).count());
+                realm.close();
+                closedLatch.countDown();
+            }
+        });
+        thread.start();
+
+        TestHelper.awaitOrFail(openedLatch);
+        Thread.sleep(100);
+        realm.beginTransaction();
+        realm.commitTransaction();
+        TestHelper.awaitOrFail(changedLatch);
+        assertTrue(result1.get());
+        assertEquals(0, count1.get());
+        Thread.sleep(500);
+        populateForDistinct(realm, numberOfBlocks, numberOfObjects, false);
+        TestHelper.awaitOrFail(closedLatch);
+        assertTrue(result2.get());
+        assertEquals(numberOfBlocks * numberOfBlocks, count2.get());
+    }
+
+    // Test if waitForChange gets waked up by stopWaitForChange called.
+    @Test
+    public void stopWaitForChange() throws InterruptedException {
+        final CountDownLatch openedLatch = new CountDownLatch(1);
+        final CountDownLatch canceledLatch = new CountDownLatch(1);
+        final CountDownLatch closedLatch = new CountDownLatch(1);
+        final AtomicBoolean result = new AtomicBoolean(false);
+        final DynamicRealm[] otherRealm = new DynamicRealm[1];
+
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                otherRealm[0] = DynamicRealm.getInstance(defaultConfig);
+                openedLatch.countDown();
+                result.set(otherRealm[0].waitForChange());
+                canceledLatch.countDown();
+                otherRealm[0].close();
+                closedLatch.countDown();
+            }
+        }).start();
+
+        TestHelper.awaitOrFail(openedLatch);
+        Thread.sleep(200);
+        otherRealm[0].stopWaitForChange();
+        TestHelper.awaitOrFail(canceledLatch);
+        assertFalse(result.get());
+        TestHelper.awaitOrFail(closedLatch);
+    }
+
+    // Test if waitForChange still blocks if stopWaitForChange has been called before.
+    @Test
+    public void waitForChange_againAfterStop() throws InterruptedException {
+        final CountDownLatch openedLatch = new CountDownLatch(1);
+        final CountDownLatch changedLatch = new CountDownLatch(1);
+        final CountDownLatch closedLatch = new CountDownLatch(1);
+        final AtomicBoolean result1 = new AtomicBoolean(true);
+        final AtomicBoolean result2 = new AtomicBoolean(false);
+        final DynamicRealm[] realmArray = new DynamicRealm[1];
```

Comment:
Why not use an AtomicReference instead? This pattern of creating 1-size arrays really decrease readability IMO and we should move away from it. This is also why there now is an ExceptionHolder class.